wolfcrypt/src/port/maxim/max3266x.c

```diff
@@ -0,0 +1,896 @@
+    mp_digit *zero_tmp = XMALLOC(multiplier->size*(sizeof(mp_digit)), NULL,
+                                    DYNAMIC_TYPE_TMP_BUFFER);
+    XMEMSET(zero_tmp, 0x00, multiplier->size*(sizeof(mp_digit)));
+
+    /* Check for invalid arguments befor padding */
+    switch(clc){
+        case(WC_MXC_TPU_MAA_EXP):
+            /* Cannot be 0 for a^e mod m operation */
+            if(XMEMCMP(zero_tmp, exp, (exp->used*sizeof(mp_digit))) == 0){
+                XFREE(zero_tmp, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                WOLFSSL_MSG("Cannot use Value 0 for Exp");
+                return BAD_FUNC_ARG;
+                break;
+            }
+
+            /* Padd out rest of data if used != length to ensure no */
+            /* garbage is used in calculation */
+            if ((exp != NULL) && (clc == WC_MXC_TPU_MAA_EXP)){
+                if((exp->dp != NULL) && (exp->used < length)){
+                    WOLFSSL_MSG("Zero Padding Exp Buffer");
+                    XMEMSET(exp->dp + exp->used, 0x00,
+                            sizeof(int) *(length - exp->used));
+                }
+            }
+
+        /* Fall through to check mod is not 0 */
+        case(WC_MXC_TPU_MAA_SQ):
+        case(WC_MXC_TPU_MAA_MUL):
+        case(WC_MXC_TPU_MAA_SQMUL):
+        case(WC_MXC_TPU_MAA_ADD):
+        case(WC_MXC_TPU_MAA_SUB):
+            /* Cannot be 0 for mod m value */
+            if(XMEMCMP(zero_tmp, mod, (exp->used*sizeof(mp_digit))) == 0){
+                XFREE(zero_tmp, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                WOLFSSL_MSG("Cannot use Value 0 for Exp");
+                return BAD_FUNC_ARG;
+                break;
+            }
+
+            /* Padd out rest of data if used != length to ensure no */
+            /* garbage is used in calculation */
+            if((multiplier->dp != NULL) && (multiplier->used < length)){
+                WOLFSSL_MSG("Zero Padding Multipler Buffer");
+                XMEMSET(multiplier->dp + multiplier->used, 0x00,
+                    sizeof(int) * (length - multiplier->used));
+            }
+            if((multiplicand->dp != NULL) && (multiplicand->used < length))
+            {
+                WOLFSSL_MSG("Zero Padding Multiplicand Buffer");
+                XMEMSET(multiplicand->dp + multiplicand->used, 0x00,
+                    sizeof(int) * (length - multiplicand->used));
+            }
+            if((mod->dp != NULL) && (mod->used < length)){
+                WOLFSSL_MSG("Zero Padding Mod Buffer");
+                XMEMSET(mod->dp + mod->used, 0x00,
+                            sizeof(int) *(length - mod->used));
+            }
+            break;
+        default:
+            return BAD_FUNC_ARG; /* Invalid clc given */
+    }
+    /* Free the zero array used to check values */
+    XFREE(zero_tmp, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+
+    /* Make sure result is 0 padded */
+    if((result->dp != NULL)){
+        XMEMSET(result->dp, 0x00, sizeof(int)*(length));
```

Comment:
Consider using ForceZero if this may contain private data... Some compilers will optimize away the memset at the end of the function if they know it is not used.